src/wix/WixToolset.Core/Link/AddDefaultSymbolsCommand.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.
+
+namespace WixToolset.Core.Link
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using WixToolset.Data;
+    using WixToolset.Data.Symbols;
+
+    internal class AddDefaultSymbolsCommand
+    {
+        public static readonly string WixStandardInstallFolder = "INSTALLFOLDER";
+        public static readonly string WixStandardInstallFolderParent = "ProgramFiles6432Folder";
+        public static readonly string WixStandardInstallFolderReference = "Directory:INSTALLFOLDER";
+
+        public AddDefaultSymbolsCommand(FindEntrySectionAndLoadSymbolsCommand find, List<IntermediateSection> sections)
+        {
+            this.Find = find;
+            this.Sections = sections;
+        }
+
+        public IntermediateSection EntrySection { get; }
+
+        public IEnumerable<IntermediateSection> Sections { get; }
+
+        public FindEntrySectionAndLoadSymbolsCommand Find { get; }
+
+        public void Execute()
+        {
+            if (this.Find.EntrySection.Type != SectionType.Package)
+            {
+                // Only packages...for now.
+                return;
+            }
+
+            var symbols = this.Sections.SelectMany(s => s.Symbols);
+            var referenceSymbols = symbols.OfType<WixSimpleReferenceSymbol>();
+            var directorySymbols = symbols.OfType<DirectorySymbol>();
+
+            if (referenceSymbols.Any(s => s.SymbolicName == WixStandardInstallFolderReference)
+                && !directorySymbols.Any(d => d.Id.Id == WixStandardInstallFolder))
+            {
+                // If there are any INSTALLFOLDER references, add a default one, using the
+                // first reference as the "canonical" reference for source line numbers.
+                this.AddSymbol(new DirectorySymbol(null, new Identifier(AccessModifier.Global, WixStandardInstallFolder))
+                {
+                    ParentDirectoryRef = WixStandardInstallFolderParent,
+                    Name = "!(bind.Property.Manufacturer) !(bind.Property.ProductName)",
+                    SourceName = ".",
+                });
+
+                this.AddSymbol(new WixSimpleReferenceSymbol(null, new Identifier(AccessModifier.Global, WixStandardInstallFolder))
+                {
+                    Table = "Directory",
+                    PrimaryKeys = WixStandardInstallFolderParent,
+                });
+            }
+
+            var upgradeSymbols = symbols.OfType<UpgradeSymbol>();
+            if (!upgradeSymbols.Any())
+            {
+                var packageSymbol = symbols.OfType<WixPackageSymbol>().SingleOrDefault();
```

Comment:
`SingleOrDefault()` will force the enumeration to go through all symbols, which could be slightly expensive for really big builds. I would use `FirstOrDefault()` and let the binder deal with making sure there is a single package symbol.

Also, the `WixPackageSymbol` should be in the `EntrySection` so it could limit the search to just that section rather than all symbols. But I think the EntrySection is the first in the list of sections so a `FirstOrDefault()` across all sections will probably perform the same.